### PR TITLE
docs: RFC on C++17 minimum standard support

### PR DESCRIPTION
**Pull request by nilason**
Addresses  #4040.

For reference:
[GDAL RFC98](https://gdal.org/en/latest/development/rfc/rfc98_build_requirements_gdal_3_9.html)
[PDAL 2.4](https://github.com/PDAL/PDAL/releases/tag/2.4.0)
[GEOS#1144](https://github.com/libgeos/geos/pull/1144)

### CODE REVIEW

**Comment by echoix** [comment]
Is there a way to write it such that the RFC doesn't have to be updated at each version? Such as the conditions that allow us to say: these are the supported ones without having to think more about it?

For Python, the case is quite easy.

**Comment by nilason** [comment]
> Is there a way to write it such that the RFC doesn't have to be updated at each version? Such as the conditions that allow us to say: these are the supported ones without having to think more about it?
> 
> For Python, the case is quite easy.

I can only repeat what I said before on this matter: https://github.com/OSGeo/grass/issues/4040#issuecomment-2230309429.

**Comment by wenzeslaus** [comment]
@echoix I would prefer the Python way as well, but I'm afraid @nilason is right that it is just more complicated for C and C++. Maybe it will change it in the future, but the nature of the ecosystem is just very different now.

One thing which we could do differently is updating the same RFC instead of creating a new one. I don't know if that's a good idea here or in general. For things like release schedule it seems to make sense, though.

**Comment by nilason** [comment]
It is probable we will want to update C and C++ minimum standard only every 5–10 years,  so a new RFC weighing in all factors is not such a big deal I think. The current standard should be mentioned in doc/development/style_guide.md.

For reference eg. GDAL RFCs work in the same way, without "updating". Each RFC is a document in itself, with its history and adoption procedure.

**Comment by wenzeslaus** [comment]
@nilason, do you want to make any changes to tested C and C++ version in relation to this? The gcc.yml workflow is already using `gnu17` and `c++17`, but perhaps adding new versions? I don't know what GCC is there, but the underlying image is currently `ubuntu-22.04`. Related to that, do you want to run the CI with clang? macOS seems like a good image to do that right now.

File: doc/development/rfc/RFC10_cpp_language_standards_support.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+regarding C++ standard support.
+
+## Background
+
+The RFC 7, which set the C++ minimal standard support to C++11, was adopted 3.5
+years ago. At that time, although the latest versions of common compilers had
+full support for C++17, GRASS GIS code itself nor its dependencies were not in
+need or required it. Now, compilers with full C++17 support are available also
+in what may be considered stable or long-term-support systems.
+Moreover, important dependencies such as PDAL 2.4 (released in Mars 2022) and
```
Comment (wenzeslaus) [inline doc/development/rfc/RFC10_cpp_language_standards_support.md:20]:
```suggestion
Moreover, important dependencies such as PDAL 2.4 (released in March 2022) and
```

File: doc/development/rfc/RFC10_cpp_language_standards_support.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+GDAL 3.9 (May 2024) require C++17 support, and so will the future release of
+GEOS 3.14 do.
```
Comment (wenzeslaus) [inline doc/development/rfc/RFC10_cpp_language_standards_support.md:22]:
```suggestion
GEOS 3.14.
```

File: doc/development/rfc/RFC10_cpp_language_standards_support.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+Author of the first draft: Nicklas Larsson
+
+Status: Draft
+
+## Summary
+
+Set new minimal build requirement for C++ code to support C++17 standard. This
+supersedes the [RFC 7: Language Standards Support](language_standards_support.md),
+regarding C++ standard support.
```
Comment (wenzeslaus) [inline doc/development/rfc/RFC10_cpp_language_standards_support.md:11]:
```suggestion
Set a new minimum build requirement for C++ code to support the C++17 standard. This
supersedes the [RFC 7: Language Standards Support](language_standards_support.md) with respect to C++ standard support.
```

File: doc/development/rfc/RFC10_cpp_language_standards_support.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+## Background
+
+The RFC 7, which set the C++ minimal standard support to C++11, was adopted 3.5
+years ago. At that time, although the latest versions of common compilers had
+full support for C++17, GRASS GIS code itself nor its dependencies were not in
+need or required it. Now, compilers with full C++17 support are available also
```
Comment (wenzeslaus) [inline doc/development/rfc/RFC10_cpp_language_standards_support.md:18]:
```suggestion
The RFC 7, which set the minimum supported C++ standard to C++11, was adopted 3.5
years ago. At that time, although the latest versions of common compilers had
full support for C++17, neither the GRASS GIS code itself nor its dependencies
needed or required it. Now, compilers with full C++17 support are available also
```

**Review by wenzeslaus — CHANGES_REQUESTED** [review]
I have some wording suggestions, but otherwise we are only testing C++17 in gcc.yml and our dependencies already require C++17, so I don't think there is much to discuss.

**Comment by nilason** [comment]
> I have some wording suggestions, but otherwise we are only testing C++17 in gcc.yml and our dependencies already require C++17, so I don't think there is much to discuss.

Thanks for the good suggestions. In practice, also the mac runner is C++17, as that is the default with Clang 19.

**Review by wenzeslaus — APPROVED** [review]
If I recall correctly, the procedure is that the PSC needs to vote here by approving the PR. Please, check PSC mailing list for the last RFC voting.

I guess we could merge the draft and then only PSC-vote to "un-draft" it. I don't see much advantage in that.

File: doc/development/rfc/RFC10_cpp_language_standards_support.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+needed or required it. Now, compilers with full C++17 support are available also
+in what may be considered stable or long-term-support systems.
+Moreover, important dependencies such as PDAL 2.4 (released in March 2022) and
+GDAL 3.9 (May 2024) require C++17 support, and so will the future release of
+GEOS 3.14.
+
+## C++17 standard for GRASS GIS
+
+The time has come to increase the minimal C++ standard support for GRASS GIS
+code to the C++17 standard. This enables use of new C++17 features, if and when
```
Comment (veroandreo) [inline doc/development/rfc/RFC10_cpp_language_standards_support.md:26]:
```suggestion
code to the C++17 standard. This enables the use of new C++17 features, if and when
```

File: doc/development/rfc/RFC10_cpp_language_standards_support.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+so is needed or recommended. There is, however, no need to make any immediate
+changes to current code. Continuous integration runners already compiles in
```
Comment (veroandreo) [inline doc/development/rfc/RFC10_cpp_language_standards_support.md:28]:
```suggestion
changes to current code. Continuous integration runners already compile in
```

**Review by veroandreo — APPROVED** [review]
Thanks @nilason!

**Review by veroandreo — DISMISSED** [review]

**Comment by echoix** [comment]
Since the motion was passed, it needs to be updated to no longer be a draft

**Comment by nilason** [comment]
I updated and activated auto-merge.

**Review by echoix — APPROVED** [review]
Approving since adopted by the RFC

**Pull request merged**